// expo-cli/build/commands/diagnostics.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var action = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee(options) {
    var info;
    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (_envinfo || _load_envinfo()).default.run({
              System: ['OS', 'Shell'],
              Binaries: ['Node', 'Yarn', 'npm', 'Watchman'],
              IDEs: ['Xcode', 'Android Studio'],
              npmPackages: ['expo', 'react', 'react-native', 'react-navigation'],
              npmGlobalPackages: ['expo-cli']
            }, {
              title: 'Expo CLI ' + (_package || _load_package()).version + ' environment info'
            });

          case 2:
            info = _context.sent;

            console.log(info);

          case 4:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function action(_x) {
    return _ref.apply(this, arguments);
  };
}();

var _envinfo;

function _load_envinfo() {
  return _envinfo = _interopRequireDefault(require('envinfo'));
}

var _package;

function _load_package() {
  return _package = require('../../package.json');
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (program) {
  program.command('diagnostics [project-dir]').description('Prints environment info to console.').asyncAction(action);
};

module.exports = exports['default'];
