// expo-cli/build/commands/start.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getIterator2;

function _load_getIterator() {
  return _getIterator2 = _interopRequireDefault(require('babel-runtime/core-js/get-iterator'));
}

var _keys;

function _load_keys() {
  return _keys = _interopRequireDefault(require('babel-runtime/core-js/object/keys'));
}

var _taggedTemplateLiteral2;

function _load_taggedTemplateLiteral() {
  return _taggedTemplateLiteral2 = _interopRequireDefault(require('babel-runtime/helpers/taggedTemplateLiteral'));
}

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var parseStartOptionsAsync = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee(projectDir, options) {
    var startOpts;
    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            startOpts = {};


            if (options.clear) {
              startOpts.reset = true;
            }

            if (options.parent && options.parent.nonInteractive) {
              startOpts.nonInteractive = true;
            }

            if (options.maxWorkers) {
              startOpts.maxWorkers = options.maxWorkers;
            }

            if (!options.webOnly) {
              _context.next = 8;
              break;
            }

            startOpts.webOnly = options.webOnly;
            _context.next = 11;
            break;

          case 8:
            _context.next = 10;
            return (_xdl || _load_xdl()).Web.onlySupportsWebAsync(projectDir);

          case 10:
            startOpts.webOnly = _context.sent;

          case 11:
            return _context.abrupt('return', startOpts);

          case 12:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function parseStartOptionsAsync(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

var action = function () {
  var _ref2 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2(projectDir, options) {
    var rootPath, devToolsUrl, _ref3, exp, pkg, nonInteractive, startOpts, url, recipient;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            (0, (_exit || _load_exit()).installExitHooks)(projectDir);

            _context2.next = 3;
            return (_urlOpts || _load_urlOpts()).default.optsAsync(projectDir, options);

          case 3:

            (0, (_log || _load_log()).default)((_chalk || _load_chalk()).default.gray('Starting project at', projectDir));

            rootPath = _path.default.resolve(projectDir);
            _context2.next = 7;
            return (_devTools || _load_devTools()).DevToolsServer.startAsync(rootPath);

          case 7:
            devToolsUrl = _context2.sent;

            (0, (_log || _load_log()).default)('Expo DevTools is running at ' + (_chalk || _load_chalk()).default.underline(devToolsUrl));

            _context2.next = 11;
            return (_xdl || _load_xdl()).ProjectUtils.readConfigJsonAsync(projectDir);

          case 11:
            _ref3 = _context2.sent;
            exp = _ref3.exp;
            pkg = _ref3.pkg;

            if (exp === null) {
              (_log || _load_log()).default.warn('No Expo configuration found. Are you sure this is a project directory?');
              process.exit(1);
            }

            _context2.next = 17;
            return validateDependenciesVersions(projectDir, exp, pkg);

          case 17:
            nonInteractive = options.parent && options.parent.nonInteractive;

            if (!(!nonInteractive && !exp.isDetached)) {
              _context2.next = 27;
              break;
            }

            _context2.next = 21;
            return (_xdl || _load_xdl()).UserSettings.getAsync('openDevToolsAtStartup', true);

          case 21:
            if (!_context2.sent) {
              _context2.next = 26;
              break;
            }

            (0, (_log || _load_log()).default)('Opening DevTools in the browser... (press ' + (_chalk || _load_chalk()).default.bold(_templateObject) + ' to disable)');
            (0, (_openBrowser || _load_openBrowser()).default)(devToolsUrl);
            _context2.next = 27;
            break;

          case 26:
            (0, (_log || _load_log()).default)('Press ' + (_chalk || _load_chalk()).default.bold(_templateObject2) + ' to open DevTools now, or ' + (_chalk || _load_chalk()).default.bold(_templateObject) + ' to always open it automatically.');

          case 27:
            _context2.next = 29;
            return parseStartOptionsAsync(projectDir, options);

          case 29:
            startOpts = _context2.sent;
            _context2.next = 32;
            return (_xdl || _load_xdl()).Project.startAsync(rootPath, startOpts);

          case 32:
            _context2.next = 34;
            return (_xdl || _load_xdl()).UrlUtils.constructManifestUrlAsync(projectDir);

          case 34:
            url = _context2.sent;
            _context2.next = 37;
            return (_sendTo || _load_sendTo()).default.getRecipient(options.sendTo);

          case 37:
            recipient = _context2.sent;

            if (!recipient) {
              _context2.next = 41;
              break;
            }

            _context2.next = 41;
            return (_sendTo || _load_sendTo()).default.sendUrlAsync(url, recipient);

          case 41:
            _context2.next = 43;
            return (_urlOpts || _load_urlOpts()).default.handleMobileOptsAsync(projectDir, options);

          case 43:
            if (!(!nonInteractive && !exp.isDetached)) {
              _context2.next = 48;
              break;
            }

            _context2.next = 46;
            return (_TerminalUI || _load_TerminalUI()).startAsync(projectDir, startOpts);

          case 46:
            _context2.next = 49;
            break;

          case 48:
            if (!options.webOnly) {
              if (!exp.isDetached) {
                (_log || _load_log()).default.newLine();
                (_urlOpts || _load_urlOpts()).default.printQRCode(url);
              }
              (0, (_log || _load_log()).default)('Your native app is running at ' + (_chalk || _load_chalk()).default.underline(url));
            }

          case 49:
            (_log || _load_log()).default.nested((_chalk || _load_chalk()).default.green('Logs for your project will appear below. Press Ctrl+C to exit.'));

          case 50:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function action(_x3, _x4) {
    return _ref2.apply(this, arguments);
  };
}();

var validateDependenciesVersions = function () {
  var _ref4 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3(projectDir, exp, pkg) {
    var bundleNativeModulesPath, bundledNativeModules, bundledNativeModulesNames, projectDependencies, modulesToCheck, incorrectDeps, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, moduleName, expectedRange, actualRange;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            if ((_xdl || _load_xdl()).Versions.gteSdkVersion(exp, '33.0.0')) {
              _context3.next = 2;
              break;
            }

            return _context3.abrupt('return');

          case 2:
            bundleNativeModulesPath = (0, (_attempt || _load_attempt()).default)(function () {
              return (_config || _load_config()).resolveModule('expo/bundledNativeModules.json', projectDir, exp);
            });

            if (!(0, (_isError || _load_isError()).default)(bundleNativeModulesPath)) {
              _context3.next = 6;
              break;
            }

            (_log || _load_log()).default.warn('Your project is in SDK version >= 33.0.0, but the ' + (_chalk || _load_chalk()).default.underline('expo') + ' package version seems to be older.');
            return _context3.abrupt('return');

          case 6:
            _context3.next = 8;
            return (_jsonFile || _load_jsonFile()).default.readAsync(bundleNativeModulesPath);

          case 8:
            bundledNativeModules = _context3.sent;
            bundledNativeModulesNames = (0, (_keys || _load_keys()).default)(bundledNativeModules);
            projectDependencies = (0, (_keys || _load_keys()).default)(pkg.dependencies);
            modulesToCheck = (0, (_intersection || _load_intersection()).default)(bundledNativeModulesNames, projectDependencies);
            incorrectDeps = [];
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            _context3.prev = 16;

            for (_iterator = (0, (_getIterator2 || _load_getIterator()).default)(modulesToCheck); !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
              moduleName = _step.value;
              expectedRange = bundledNativeModules[moduleName];
              actualRange = pkg.dependencies[moduleName];

              if (!(_semver || _load_semver()).default.intersects(expectedRange, actualRange)) {
                incorrectDeps.push({
                  moduleName: moduleName,
                  expectedRange: expectedRange,
                  actualRange: actualRange
                });
              }
            }
            _context3.next = 24;
            break;

          case 20:
            _context3.prev = 20;
            _context3.t0 = _context3['catch'](16);
            _didIteratorError = true;
            _iteratorError = _context3.t0;

          case 24:
            _context3.prev = 24;
            _context3.prev = 25;

            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }

          case 27:
            _context3.prev = 27;

            if (!_didIteratorError) {
              _context3.next = 30;
              break;
            }

            throw _iteratorError;

          case 30:
            return _context3.finish(27);

          case 31:
            return _context3.finish(24);

          case 32:
            if (incorrectDeps.length > 0) {
              (_log || _load_log()).default.warn("Some of your project's dependencies are not compatible with currently installed expo package version:");
              incorrectDeps.forEach(function (_ref5) {
                var moduleName = _ref5.moduleName,
                    expectedRange = _ref5.expectedRange,
                    actualRange = _ref5.actualRange;

                (_log || _load_log()).default.warn(' - ' + (_chalk || _load_chalk()).default.underline(moduleName) + ' - expected version range: ' + (_chalk || _load_chalk()).default.underline(expectedRange) + ' - actual version installed: ' + (_chalk || _load_chalk()).default.underline(actualRange));
              });
              (_log || _load_log()).default.warn('Your project may not work correctly until you install the correct versions of the packages.\n' + ('To install the correct versions of these packages, please run: ' + (_chalk || _load_chalk()).default.inverse('expo install [package-name ...]')));
            }

          case 33:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this, [[16, 20, 24, 32], [25,, 27, 31]]);
  }));

  return function validateDependenciesVersions(_x5, _x6, _x7) {
    return _ref4.apply(this, arguments);
  };
}();

var _templateObject = (0, (_taggedTemplateLiteral2 || _load_taggedTemplateLiteral()).default)(['shift-d'], ['shift-d']),
    _templateObject2 = (0, (_taggedTemplateLiteral2 || _load_taggedTemplateLiteral()).default)(['d'], ['d']);

var _path = _interopRequireDefault(require('path'));

var _config;

function _load_config() {
  return _config = _interopRequireWildcard(require('@expo/config'));
}

var _devTools;

function _load_devTools() {
  return _devTools = require('@expo/dev-tools');
}

var _jsonFile;

function _load_jsonFile() {
  return _jsonFile = _interopRequireDefault(require('@expo/json-file'));
}

var _xdl;

function _load_xdl() {
  return _xdl = require('@expo/xdl');
}

var _chalk;

function _load_chalk() {
  return _chalk = _interopRequireDefault(require('chalk'));
}

var _openBrowser;

function _load_openBrowser() {
  return _openBrowser = _interopRequireDefault(require('react-dev-utils/openBrowser'));
}

var _intersection;

function _load_intersection() {
  return _intersection = _interopRequireDefault(require('lodash/intersection'));
}

var _semver;

function _load_semver() {
  return _semver = _interopRequireDefault(require('semver'));
}

var _attempt;

function _load_attempt() {
  return _attempt = _interopRequireDefault(require('lodash/attempt'));
}

var _isError;

function _load_isError() {
  return _isError = _interopRequireDefault(require('lodash/isError'));
}

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('../log'));
}

var _sendTo;

function _load_sendTo() {
  return _sendTo = _interopRequireDefault(require('../sendTo'));
}

var _exit;

function _load_exit() {
  return _exit = require('../exit');
}

var _urlOpts;

function _load_urlOpts() {
  return _urlOpts = _interopRequireDefault(require('../urlOpts'));
}

var _TerminalUI;

function _load_TerminalUI() {
  return _TerminalUI = _interopRequireWildcard(require('./start/TerminalUI'));
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (program) {
  program.command('start [project-dir]').alias('r').description('Starts or restarts a local server for your app and gives you a URL to it').option('-s, --send-to [dest]', 'An email address to send a link to').option('-c, --clear', 'Clear the React Native packager cache').option('--web-only', 'Only start the webpack server')
  // TODO(anp) set a default for this dynamically based on whether we're inside a container?
  .option('--max-workers [num]', 'Maximum number of tasks to allow Metro to spawn.').urlOpts().allowOffline().asyncActionProjectDir(action, true, true);
};

module.exports = exports['default'];
