// xdl/build/Sentry.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.logError = logError;
exports.captureException = captureException;

var _raven;

function _load_raven() {
  return _raven = _interopRequireDefault(require('raven'));
}

var _Config;

function _load_Config() {
  return _Config = _interopRequireDefault(require('./Config'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const SENTRY_DSN = 'https://8554f14d112d4ed4b0558154762760ef:bae5673d5e5243abac5563d70861b5d8@sentry.io/194120';

(_raven || _load_raven()).default.config(SENTRY_DSN).install();

function logError(message, options) {
  // send error to Sentry
  // add `testing: true` to tags to avoid sending an email when testing
  (_raven || _load_raven()).default.captureMessage(message, getOptions(options));
}

function captureException(ex, options) {
  (_raven || _load_raven()).default.captureException(ex, getOptions(options));
}

function getOptions(options = {}) {
  return _extends({}, options, {
    tags: _extends({}, options.tags, {
      developerTool: (_Config || _load_Config()).default.developerTool,
      offline: (_Config || _load_Config()).default.offline
    })
  });
}
