// expo-cli/build/commands/build/ios/appleApi/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setup = exports.ensureAppExists = exports.createManagers = exports.authenticate = undefined;

var _authenticate;

function _load_authenticate() {
  return _authenticate = _interopRequireDefault(require('./authenticate'));
}

var _distributionCert;

function _load_distributionCert() {
  return _distributionCert = _interopRequireDefault(require('./distributionCert'));
}

var _pushKey;

function _load_pushKey() {
  return _pushKey = _interopRequireDefault(require('./pushKey'));
}

var _provisioningProfile;

function _load_provisioningProfile() {
  return _provisioningProfile = _interopRequireDefault(require('./provisioningProfile'));
}

var _ensureAppExists;

function _load_ensureAppExists() {
  return _ensureAppExists = _interopRequireDefault(require('./ensureAppExists'));
}

var _setup;

function _load_setup() {
  return _setup = _interopRequireDefault(require('./setup'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createManagers = function createManagers(ctx) {
  return {
    distributionCert: (0, (_distributionCert || _load_distributionCert()).default)(ctx),
    pushKey: (0, (_pushKey || _load_pushKey()).default)(ctx),
    provisioningProfile: (0, (_provisioningProfile || _load_provisioningProfile()).default)(ctx)
  };
};

exports.authenticate = (_authenticate || _load_authenticate()).default;
exports.createManagers = createManagers;
exports.ensureAppExists = (_ensureAppExists || _load_ensureAppExists()).default;
exports.setup = (_setup || _load_setup()).default;
