// expo-cli/build/commands/build/ios/credentials/revoke.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _getIterator2;

function _load_getIterator() {
  return _getIterator2 = _interopRequireDefault(require('babel-runtime/core-js/get-iterator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var revoke = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee(appleCtx, typesToRevoke) {
    var managers, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, type, manager;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            managers = (_appleApi || _load_appleApi()).createManagers(appleCtx);
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            _context.prev = 4;
            _iterator = (0, (_getIterator2 || _load_getIterator()).default)(typesToRevoke);

          case 6:
            if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
              _context.next = 15;
              break;
            }

            type = _step.value;
            manager = managers[type];

            if (!manager) {
              _context.next = 12;
              break;
            }

            _context.next = 12;
            return _revokeByType(manager);

          case 12:
            _iteratorNormalCompletion = true;
            _context.next = 6;
            break;

          case 15:
            _context.next = 21;
            break;

          case 17:
            _context.prev = 17;
            _context.t0 = _context['catch'](4);
            _didIteratorError = true;
            _iteratorError = _context.t0;

          case 21:
            _context.prev = 21;
            _context.prev = 22;

            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }

          case 24:
            _context.prev = 24;

            if (!_didIteratorError) {
              _context.next = 27;
              break;
            }

            throw _iteratorError;

          case 27:
            return _context.finish(24);

          case 28:
            return _context.finish(21);

          case 29:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[4, 17, 21, 29], [22,, 24, 28]]);
  }));

  return function revoke(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

var _revokeByType = function () {
  var _ref2 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2(manager) {
    var list, choices, ids;
    return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            if (!manager.list) {
              _context2.next = 14;
              break;
            }

            _context2.next = 3;
            return manager.list();

          case 3:
            list = _context2.sent;

            if (!(!list || list.length === 0)) {
              _context2.next = 6;
              break;
            }

            return _context2.abrupt('return');

          case 6:
            choices = list.map(function (cert) {
              return { name: manager.format(cert), value: cert.id };
            });
            _context2.next = 9;
            return _askWhatToRevoke(choices);

          case 9:
            ids = _context2.sent;
            _context2.next = 12;
            return manager.revoke(ids);

          case 12:
            _context2.next = 16;
            break;

          case 14:
            _context2.next = 16;
            return manager.revoke();

          case 16:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function _revokeByType(_x3) {
    return _ref2.apply(this, arguments);
  };
}();

var _askWhatToRevoke = function () {
  var _ref3 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3(choices) {
    var _ref4, ids;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return (0, (_prompt || _load_prompt()).default)({
              type: 'checkbox',
              name: 'ids',
              message: 'What would you like to revoke?',
              pageSize: Infinity,
              choices: choices
            });

          case 2:
            _ref4 = _context3.sent;
            ids = _ref4.ids;
            return _context3.abrupt('return', ids);

          case 5:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this);
  }));

  return function _askWhatToRevoke(_x4) {
    return _ref3.apply(this, arguments);
  };
}();

var _appleApi;

function _load_appleApi() {
  return _appleApi = _interopRequireWildcard(require('../appleApi'));
}

var _prompt;

function _load_prompt() {
  return _prompt = _interopRequireDefault(require('../../../../prompt'));
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = revoke;
module.exports = exports['default'];
