// kinvey-js-sdk/lib/endpoint.js
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var isString_1 = tslib_1.__importDefault(require("lodash/isString"));
var http_1 = require("./http");
var kinvey_1 = require("./errors/kinvey");
function endpoint(endpoint, args, options) {
    if (options === void 0) { options = {}; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var request, response;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!isString_1.default(endpoint)) {
                        throw new kinvey_1.KinveyError('An endpoint is required and must be a string.');
                    }
                    request = new http_1.KinveyHttpRequest({
                        method: http_1.HttpRequestMethod.POST,
                        auth: http_1.KinveyHttpAuth.SessionOrMaster,
                        url: http_1.formatKinveyBaasUrl(http_1.KinveyBaasNamespace.Rpc, "/custom/" + endpoint),
                        body: args,
                        timeout: options.timeout
                    });
                    return [4 /*yield*/, request.execute()];
                case 1:
                    response = _a.sent();
                    return [2 /*return*/, response.data];
            }
        });
    });
}
exports.endpoint = endpoint;
