// kinvey-js-sdk/lib/storage.js
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var isArray_1 = tslib_1.__importDefault(require("lodash/isArray"));
var p_queue_1 = tslib_1.__importDefault(require("p-queue"));
var config_1 = require("./config");
var query_1 = require("./query");
var aggregation_1 = require("./aggregation");
var kinvey_1 = require("./errors/kinvey");
var http_1 = require("./http");
var queue = new p_queue_1.default({ concurrency: 1 });
function generateId(length) {
    if (length === void 0) { length = 24; }
    var chars = 'abcdef0123456789';
    var id = '';
    for (var i = 0, j = chars.length; i < length; i += 1) {
        var pos = Math.floor(Math.random() * j);
        id += chars.substring(pos, pos + 1);
    }
    return id;
}
var Storage = /** @class */ (function () {
    function Storage(dbName, collectionName) {
        this.dbName = dbName;
        this.collectionName = collectionName;
    }
    Object.defineProperty(Storage.prototype, "storageAdapter", {
        get: function () {
            return config_1.getConfig(config_1.ConfigKey.StorageAdapter);
        },
        enumerable: true,
        configurable: true
    });
    Storage.prototype.find = function (query) {
        var _this = this;
        return queue.add(function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var docs;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (query && !(query instanceof query_1.Query)) {
                            throw new kinvey_1.KinveyError('Invalid query. It must be an instance of the Query class.');
                        }
                        return [4 /*yield*/, this.storageAdapter.find(this.dbName, this.collectionName)];
                    case 1:
                        docs = _a.sent();
                        if (docs.length > 0 && query) {
                            return [2 /*return*/, query.process(docs)];
                        }
                        return [2 /*return*/, docs];
                }
            });
        }); });
    };
    Storage.prototype.group = function (aggregation) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var docs;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(aggregation instanceof aggregation_1.Aggregation)) {
                            throw new kinvey_1.KinveyError('Invalid aggregation. It must be an instance of the Aggregation class.');
                        }
                        return [4 /*yield*/, this.find()];
                    case 1:
                        docs = _a.sent();
                        return [2 /*return*/, aggregation.process(docs)];
                }
            });
        });
    };
    Storage.prototype.count = function (query) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var docs;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!query) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.find(query)];
                    case 1:
                        docs = _a.sent();
                        return [2 /*return*/, docs.length];
                    case 2: return [2 /*return*/, queue.add(function () { return _this.storageAdapter.count(_this.dbName, _this.collectionName); })];
                }
            });
        });
    };
    Storage.prototype.findById = function (id) {
        var _this = this;
        return queue.add(function () { return _this.storageAdapter.findById(_this.dbName, _this.collectionName, id); });
    };
    Storage.prototype.save = function (docs) {
        var _this = this;
        return queue.add(function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var docsToSave, singular;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        docsToSave = docs;
                        singular = false;
                        if (!docs) {
                            return [2 /*return*/, null];
                        }
                        if (!isArray_1.default(docs)) {
                            singular = true;
                            docsToSave = [docs];
                        }
                        // Clone the docs
                        docsToSave = docsToSave.slice(0, docsToSave.length);
                        if (!(docsToSave.length > 0)) return [3 /*break*/, 2];
                        docsToSave = docsToSave.map(function (doc) {
                            if (!doc._id) {
                                return Object.assign({}, {
                                    _id: generateId(),
                                    _kmd: Object.assign({}, doc._kmd, { local: true })
                                }, doc);
                            }
                            return doc;
                        });
                        return [4 /*yield*/, this.storageAdapter.save(this.dbName, this.collectionName, docsToSave)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, singular ? docsToSave.shift() : docsToSave];
                    case 2: return [2 /*return*/, docs];
                }
            });
        }); });
    };
    Storage.prototype.remove = function (query) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var docs, results;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.find(query)];
                    case 1:
                        docs = _a.sent();
                        if (!query) return [3 /*break*/, 3];
                        return [4 /*yield*/, Promise.all(docs.map(function (doc) { return _this.removeById(doc._id); }))];
                    case 2:
                        results = _a.sent();
                        return [2 /*return*/, results.reduce(function (totalCount, count) { return totalCount + count; }, 0)];
                    case 3: return [4 /*yield*/, this.clear()];
                    case 4:
                        _a.sent();
                        return [2 /*return*/, docs.length];
                }
            });
        });
    };
    Storage.prototype.removeById = function (id) {
        var _this = this;
        return queue.add(function () { return _this.storageAdapter.removeById(_this.dbName, _this.collectionName, id); });
    };
    Storage.prototype.clear = function () {
        var _this = this;
        return queue.add(function () { return _this.storageAdapter.clear(_this.dbName, _this.collectionName); });
    };
    Storage.clear = function (dbName) {
        var storageAdapter = config_1.getConfig(config_1.ConfigKey.StorageAdapter);
        var exclude = [http_1.getKey()];
        return queue.add(function () { return storageAdapter.clearDatabase(dbName, exclude); });
    };
    return Storage;
}());
exports.Storage = Storage;
