// expo-cli/build/PackageManager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.YarnPackageManager = exports.NpmPackageManager = undefined;

var _toConsumableArray2;

function _load_toConsumableArray() {
  return _toConsumableArray2 = _interopRequireDefault(require('babel-runtime/helpers/toConsumableArray'));
}

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var _getPrototypeOf;

function _load_getPrototypeOf() {
  return _getPrototypeOf = _interopRequireDefault(require('babel-runtime/core-js/object/get-prototype-of'));
}

var _classCallCheck2;

function _load_classCallCheck() {
  return _classCallCheck2 = _interopRequireDefault(require('babel-runtime/helpers/classCallCheck'));
}

var _createClass2;

function _load_createClass() {
  return _createClass2 = _interopRequireDefault(require('babel-runtime/helpers/createClass'));
}

var _possibleConstructorReturn2;

function _load_possibleConstructorReturn() {
  return _possibleConstructorReturn2 = _interopRequireDefault(require('babel-runtime/helpers/possibleConstructorReturn'));
}

var _inherits2;

function _load_inherits() {
  return _inherits2 = _interopRequireDefault(require('babel-runtime/helpers/inherits'));
}

exports.createForProject = createForProject;

var _ansiRegex;

function _load_ansiRegex() {
  return _ansiRegex = _interopRequireDefault(require('ansi-regex'));
}

var _config;

function _load_config() {
  return _config = require('@expo/config');
}

var _spawnAsync;

function _load_spawnAsync() {
  return _spawnAsync = _interopRequireDefault(require('@expo/spawn-async'));
}

var _split;

function _load_split() {
  return _split = _interopRequireDefault(require('split'));
}

var _stream = require('stream');

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('./log'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ansi = '(?:' + (0, (_ansiRegex || _load_ansiRegex()).default)().source + ')*';
var npmPeerDependencyWarningPattern = new RegExp(ansi + 'npm' + ansi + ' ' + ansi + 'WARN' + ansi + '.+You must install peer dependencies yourself\\.\n', 'g');
var yarnPeerDependencyWarningPattern = new RegExp(ansi + 'warning' + ansi + ' "[^"]+" has (?:unmet|incorrect) peer dependency "[^"]+"\\.\n', 'g');

var NpmStderrTransform = function (_Transform) {
  (0, (_inherits2 || _load_inherits()).default)(NpmStderrTransform, _Transform);

  function NpmStderrTransform() {
    (0, (_classCallCheck2 || _load_classCallCheck()).default)(this, NpmStderrTransform);
    return (0, (_possibleConstructorReturn2 || _load_possibleConstructorReturn()).default)(this, (NpmStderrTransform.__proto__ || (0, (_getPrototypeOf || _load_getPrototypeOf()).default)(NpmStderrTransform)).apply(this, arguments));
  }

  (0, (_createClass2 || _load_createClass()).default)(NpmStderrTransform, [{
    key: '_transform',
    value: function _transform(chunk, encoding, callback) {
      this.push(chunk.toString().replace(npmPeerDependencyWarningPattern, ''));
      callback();
    }
  }]);
  return NpmStderrTransform;
}(_stream.Transform);

var YarnStderrTransform = function (_Transform2) {
  (0, (_inherits2 || _load_inherits()).default)(YarnStderrTransform, _Transform2);

  function YarnStderrTransform() {
    (0, (_classCallCheck2 || _load_classCallCheck()).default)(this, YarnStderrTransform);
    return (0, (_possibleConstructorReturn2 || _load_possibleConstructorReturn()).default)(this, (YarnStderrTransform.__proto__ || (0, (_getPrototypeOf || _load_getPrototypeOf()).default)(YarnStderrTransform)).apply(this, arguments));
  }

  (0, (_createClass2 || _load_createClass()).default)(YarnStderrTransform, [{
    key: '_transform',
    value: function _transform(chunk, encoding, callback) {
      this.push(chunk.toString().replace(yarnPeerDependencyWarningPattern, ''));
      callback();
    }
  }]);
  return YarnStderrTransform;
}(_stream.Transform);

var NpmPackageManager = exports.NpmPackageManager = function () {
  function NpmPackageManager(_ref) {
    var cwd = _ref.cwd;
    (0, (_classCallCheck2 || _load_classCallCheck()).default)(this, NpmPackageManager);

    this.options = { cwd: cwd, stdio: ['inherit', 'inherit', 'pipe'] };
  }

  (0, (_createClass2 || _load_createClass()).default)(NpmPackageManager, [{
    key: 'installAsync',
    value: function () {
      var _ref2 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee() {
        return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this._runAsync(['install']);

              case 2:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function installAsync() {
        return _ref2.apply(this, arguments);
      }

      return installAsync;
    }()
  }, {
    key: 'addAsync',
    value: function () {
      var _ref3 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2() {
        for (var _len = arguments.length, names = Array(_len), _key = 0; _key < _len; _key++) {
          names[_key] = arguments[_key];
        }

        return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._runAsync(['install', '--save'].concat(names));

              case 2:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function addAsync() {
        return _ref3.apply(this, arguments);
      }

      return addAsync;
    }()
  }, {
    key: 'addDevAsync',
    value: function () {
      var _ref4 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3() {
        for (var _len2 = arguments.length, names = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
          names[_key2] = arguments[_key2];
        }

        return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._runAsync(['install', '--save-dev'].concat(names));

              case 2:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function addDevAsync() {
        return _ref4.apply(this, arguments);
      }

      return addDevAsync;
    }()

    // Private

  }, {
    key: '_runAsync',
    value: function () {
      var _ref5 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee4(args) {
        var promise;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                (0, (_log || _load_log()).default)('> npm ' + args.join(' '));
                promise = (0, (_spawnAsync || _load_spawnAsync()).default)('npm', [].concat((0, (_toConsumableArray2 || _load_toConsumableArray()).default)(args)), this.options);

                promise.child.stderr.pipe((0, (_split || _load_split()).default)(/\r?\n/, function (line) {
                  return line + '\n';
                })).pipe(new NpmStderrTransform()).pipe(process.stderr);
                _context4.next = 5;
                return promise;

              case 5:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function _runAsync(_x) {
        return _ref5.apply(this, arguments);
      }

      return _runAsync;
    }()
  }, {
    key: 'name',
    get: function get() {
      return 'npm';
    }
  }]);
  return NpmPackageManager;
}();

var YarnPackageManager = exports.YarnPackageManager = function () {
  function YarnPackageManager(_ref6) {
    var cwd = _ref6.cwd;
    (0, (_classCallCheck2 || _load_classCallCheck()).default)(this, YarnPackageManager);

    this.options = {
      cwd: cwd,
      stdio: ['inherit', 'inherit', 'pipe']
    };
  }

  (0, (_createClass2 || _load_createClass()).default)(YarnPackageManager, [{
    key: 'installAsync',
    value: function () {
      var _ref7 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee5() {
        return (_regenerator || _load_regenerator()).default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._runAsync(['install']);

              case 2:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function installAsync() {
        return _ref7.apply(this, arguments);
      }

      return installAsync;
    }()
  }, {
    key: 'addAsync',
    value: function () {
      var _ref8 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee6() {
        for (var _len3 = arguments.length, names = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
          names[_key3] = arguments[_key3];
        }

        return (_regenerator || _load_regenerator()).default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._runAsync(['add'].concat(names));

              case 2:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function addAsync() {
        return _ref8.apply(this, arguments);
      }

      return addAsync;
    }()
  }, {
    key: 'addDevAsync',
    value: function () {
      var _ref9 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee7() {
        for (var _len4 = arguments.length, names = Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
          names[_key4] = arguments[_key4];
        }

        return (_regenerator || _load_regenerator()).default.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._runAsync(['add', '--dev'].concat(names));

              case 2:
              case 'end':
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function addDevAsync() {
        return _ref9.apply(this, arguments);
      }

      return addDevAsync;
    }()

    // Private

  }, {
    key: '_runAsync',
    value: function () {
      var _ref10 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee8(args) {
        var promise;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                (0, (_log || _load_log()).default)('> yarn ' + args.join(' '));
                promise = (0, (_spawnAsync || _load_spawnAsync()).default)('yarnpkg', args, this.options);

                promise.child.stderr.pipe(new YarnStderrTransform()).pipe(process.stderr);
                _context8.next = 5;
                return promise;

              case 5:
              case 'end':
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function _runAsync(_x2) {
        return _ref10.apply(this, arguments);
      }

      return _runAsync;
    }()
  }, {
    key: 'name',
    get: function get() {
      return 'Yarn';
    }
  }]);
  return YarnPackageManager;
}();

function createForProject(projectRoot) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  var PackageManager = void 0;
  if (options.npm) {
    PackageManager = NpmPackageManager;
  } else if (options.yarn) {
    PackageManager = YarnPackageManager;
  } else if ((0, (_config || _load_config()).isUsingYarn)(projectRoot)) {
    PackageManager = YarnPackageManager;
  } else {
    PackageManager = NpmPackageManager;
  }
  return new PackageManager({ cwd: projectRoot });
}
