// expo-cli/build/commands/upload/AndroidUploader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var _getPrototypeOf;

function _load_getPrototypeOf() {
  return _getPrototypeOf = _interopRequireDefault(require('babel-runtime/core-js/object/get-prototype-of'));
}

var _classCallCheck2;

function _load_classCallCheck() {
  return _classCallCheck2 = _interopRequireDefault(require('babel-runtime/helpers/classCallCheck'));
}

var _createClass2;

function _load_createClass() {
  return _createClass2 = _interopRequireDefault(require('babel-runtime/helpers/createClass'));
}

var _possibleConstructorReturn2;

function _load_possibleConstructorReturn() {
  return _possibleConstructorReturn2 = _interopRequireDefault(require('babel-runtime/helpers/possibleConstructorReturn'));
}

var _inherits2;

function _load_inherits() {
  return _inherits2 = _interopRequireDefault(require('babel-runtime/helpers/inherits'));
}

var _has;

function _load_has() {
  return _has = _interopRequireDefault(require('lodash/has'));
}

var _fsExtra;

function _load_fsExtra() {
  return _fsExtra = _interopRequireDefault(require('fs-extra'));
}

var _BaseUploader2;

function _load_BaseUploader() {
  return _BaseUploader2 = _interopRequireDefault(require('./BaseUploader'));
}

var _utils;

function _load_utils() {
  return _utils = require('./utils');
}

var _prompt;

function _load_prompt() {
  return _prompt = _interopRequireDefault(require('../../prompt'));
}

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('../../log'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PLATFORM = 'android';

var FILE_DOESNT_EXIST_ERROR = function FILE_DOESNT_EXIST_ERROR(path) {
  return 'File ' + path + ' doesn\'t exist.';
};

var SERVICE_ACCOUNT_JSON_QUESTION = {
  name: 'key',
  message: 'The path to the file containing service account JSON, used to authenticate with Google:',
  type: 'input',
  validate: function validate(path) {
    var done = this.async();
    (_fsExtra || _load_fsExtra()).default.pathExists(path).then(function (exists) {
      if (exists) {
        done(null, true);
      } else {
        done(FILE_DOESNT_EXIST_ERROR(path));
      }
    }).catch(function (err) {
      return done(err);
    });
  }
};

var AndroidUploader = function (_BaseUploader) {
  (0, (_inherits2 || _load_inherits()).default)(AndroidUploader, _BaseUploader);

  function AndroidUploader(projectDir, options) {
    (0, (_classCallCheck2 || _load_classCallCheck()).default)(this, AndroidUploader);
    return (0, (_possibleConstructorReturn2 || _load_possibleConstructorReturn()).default)(this, (AndroidUploader.__proto__ || (0, (_getPrototypeOf || _load_getPrototypeOf()).default)(AndroidUploader)).call(this, PLATFORM, projectDir, options));
  }

  (0, (_createClass2 || _load_createClass()).default)(AndroidUploader, [{
    key: '_ensureExperienceIsValid',
    value: function _ensureExperienceIsValid(exp) {
      if (!(0, (_has || _load_has()).default)(exp, 'android.package')) {
        throw new Error('You must specify an Android package in app.json.');
      }
    }
  }, {
    key: '_getPlatformSpecificOptions',
    value: function () {
      var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee() {
        var key;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this._getServiceAccountJSONPath();

              case 2:
                key = _context.sent;
                return _context.abrupt('return', { key: key, track: this.options.track });

              case 4:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function _getPlatformSpecificOptions() {
        return _ref.apply(this, arguments);
      }

      return _getPlatformSpecificOptions;
    }()
  }, {
    key: '_getServiceAccountJSONPath',
    value: function () {
      var _ref2 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2() {
        var key;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                key = this.options.key;
                _context2.t0 = key;

                if (!_context2.t0) {
                  _context2.next = 6;
                  break;
                }

                _context2.next = 5;
                return (_fsExtra || _load_fsExtra()).default.pathExists(key);

              case 5:
                _context2.t0 = _context2.sent;

              case 6:
                if (!_context2.t0) {
                  _context2.next = 10;
                  break;
                }

                return _context2.abrupt('return', key);

              case 10:
                if (key) {
                  (0, (_log || _load_log()).default)(FILE_DOESNT_EXIST_ERROR(key));
                }
                _context2.next = 13;
                return this._askForServiceAccountJSONPath();

              case 13:
                return _context2.abrupt('return', _context2.sent);

              case 14:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function _getServiceAccountJSONPath() {
        return _ref2.apply(this, arguments);
      }

      return _getServiceAccountJSONPath;
    }()
  }, {
    key: '_askForServiceAccountJSONPath',
    value: function () {
      var _ref3 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3() {
        var answer;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return (0, (_prompt || _load_prompt()).default)(SERVICE_ACCOUNT_JSON_QUESTION);

              case 2:
                answer = _context3.sent;
                return _context3.abrupt('return', answer.key);

              case 4:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function _askForServiceAccountJSONPath() {
        return _ref3.apply(this, arguments);
      }

      return _askForServiceAccountJSONPath;
    }()
  }, {
    key: '_uploadToTheStore',
    value: function () {
      var _ref4 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee4(platformData, path) {
        var fastlane, key, track, androidPackage;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                fastlane = this.fastlane;
                key = platformData.key, track = platformData.track;
                androidPackage = this._exp.android.package;
                _context4.next = 5;
                return (0, (_utils || _load_utils()).runFastlaneAsync)(fastlane.supplyAndroid, [path, androidPackage, key, track], {}, true);

              case 5:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function _uploadToTheStore(_x, _x2) {
        return _ref4.apply(this, arguments);
      }

      return _uploadToTheStore;
    }()
  }]);
  return AndroidUploader;
}((_BaseUploader2 || _load_BaseUploader()).default);

exports.default = AndroidUploader;
module.exports = exports['default'];
