// kinvey-js-sdk/lib/datastore/cache.js
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var isString_1 = tslib_1.__importDefault(require("lodash/isString"));
var storage_1 = require("../storage");
var kinvey_1 = require("../kinvey");
var QUERY_CACHE_TAG = '_QueryCache';
var SYNC_CACHE_TAG = 'kinvey_sync';
function isValidTag(tag) {
    var regexp = /^[a-z0-9-]+$/i;
    return isString_1.default(tag) && regexp.test(tag);
}
exports.isValidTag = isValidTag;
var DataStoreCache = /** @class */ (function (_super) {
    tslib_1.__extends(DataStoreCache, _super);
    function DataStoreCache(collectionName, tag) {
        var _this = this;
        if (tag && !isValidTag(tag)) {
            throw new Error('A tag can only contain letters, numbers, and "-".');
        }
        if (tag) {
            _this = _super.call(this, kinvey_1.getAppKey(), collectionName + "." + tag) || this;
        }
        else {
            _this = _super.call(this, kinvey_1.getAppKey(), collectionName) || this;
        }
        return _this;
    }
    DataStoreCache.clear = function () {
        return storage_1.Storage.clear(kinvey_1.getAppKey());
    };
    return DataStoreCache;
}(storage_1.Storage));
exports.DataStoreCache = DataStoreCache;
var QueryCache = /** @class */ (function (_super) {
    tslib_1.__extends(QueryCache, _super);
    function QueryCache(tag) {
        return _super.call(this, QUERY_CACHE_TAG, tag) || this;
    }
    return QueryCache;
}(DataStoreCache));
exports.QueryCache = QueryCache;
var SyncEvent;
(function (SyncEvent) {
    SyncEvent["Create"] = "POST";
    SyncEvent["Update"] = "PUT";
    SyncEvent["Delete"] = "DELETE";
})(SyncEvent = exports.SyncEvent || (exports.SyncEvent = {}));
;
var SyncCache = /** @class */ (function (_super) {
    tslib_1.__extends(SyncCache, _super);
    function SyncCache(tag) {
        return _super.call(this, SYNC_CACHE_TAG, tag) || this;
    }
    return SyncCache;
}(DataStoreCache));
exports.SyncCache = SyncCache;
