// expo-cli/build/commands/build/AndroidBuilder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var _getPrototypeOf;

function _load_getPrototypeOf() {
  return _getPrototypeOf = _interopRequireDefault(require('babel-runtime/core-js/object/get-prototype-of'));
}

var _classCallCheck2;

function _load_classCallCheck() {
  return _classCallCheck2 = _interopRequireDefault(require('babel-runtime/helpers/classCallCheck'));
}

var _createClass2;

function _load_createClass() {
  return _createClass2 = _interopRequireDefault(require('babel-runtime/helpers/createClass'));
}

var _possibleConstructorReturn2;

function _load_possibleConstructorReturn() {
  return _possibleConstructorReturn2 = _interopRequireDefault(require('babel-runtime/helpers/possibleConstructorReturn'));
}

var _inherits2;

function _load_inherits() {
  return _inherits2 = _interopRequireDefault(require('babel-runtime/helpers/inherits'));
}

var _fsExtra;

function _load_fsExtra() {
  return _fsExtra = _interopRequireDefault(require('fs-extra'));
}

var _path = _interopRequireDefault(require('path'));

var _untildify;

function _load_untildify() {
  return _untildify = _interopRequireDefault(require('untildify'));
}

var _xdl;

function _load_xdl() {
  return _xdl = require('@expo/xdl');
}

var _chalk;

function _load_chalk() {
  return _chalk = _interopRequireDefault(require('chalk'));
}

var _get;

function _load_get() {
  return _get = _interopRequireDefault(require('lodash/get'));
}

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('../../log'));
}

var _BuildError;

function _load_BuildError() {
  return _BuildError = _interopRequireDefault(require('./BuildError'));
}

var _BaseBuilder2;

function _load_BaseBuilder() {
  return _BaseBuilder2 = _interopRequireDefault(require('./BaseBuilder'));
}

var _prompt;

function _load_prompt() {
  return _prompt = _interopRequireDefault(require('../../prompt'));
}

var _utils;

function _load_utils() {
  return _utils = _interopRequireWildcard(require('./utils'));
}

var _constants;

function _load_constants() {
  return _constants = require('./constants');
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ANDROID = (_constants || _load_constants()).PLATFORMS.ANDROID;

var AndroidBuilder = function (_BaseBuilder) {
  (0, (_inherits2 || _load_inherits()).default)(AndroidBuilder, _BaseBuilder);

  function AndroidBuilder() {
    (0, (_classCallCheck2 || _load_classCallCheck()).default)(this, AndroidBuilder);
    return (0, (_possibleConstructorReturn2 || _load_possibleConstructorReturn()).default)(this, (AndroidBuilder.__proto__ || (0, (_getPrototypeOf || _load_getPrototypeOf()).default)(AndroidBuilder)).apply(this, arguments));
  }

  (0, (_createClass2 || _load_createClass()).default)(AndroidBuilder, [{
    key: 'run',
    value: function () {
      var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee() {
        var publishedExpIds;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.validateProject();

              case 2:
                _context.next = 4;
                return (_xdl || _load_xdl()).Android.checkSplashScreenImages(this.projectDir);

              case 4:
                _context.next = 6;
                return this.checkForBuildInProgress();

              case 6:
                _context.next = 8;
                return this.collectAndValidateCredentials();

              case 8:
                if (!this.options.publicUrl) {
                  _context.next = 12;
                  break;
                }

                _context.t0 = undefined;
                _context.next = 15;
                break;

              case 12:
                _context.next = 14;
                return this.ensureReleaseExists();

              case 14:
                _context.t0 = _context.sent;

              case 15:
                publishedExpIds = _context.t0;

                if (this.options.publicUrl) {
                  _context.next = 19;
                  break;
                }

                _context.next = 19;
                return this.checkStatusBeforeBuild();

              case 19:
                _context.next = 21;
                return this.build(publishedExpIds);

              case 21:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function run() {
        return _ref.apply(this, arguments);
      }

      return run;
    }()
  }, {
    key: 'validateProject',
    value: function () {
      var _ref2 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2() {
        var androidPackage;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return (_utils || _load_utils()).checkIfSdkIsSupported(this.manifest.sdkVersion, ANDROID);

              case 2:
                if ((0, (_get || _load_get()).default)(this.manifest, 'android.package')) {
                  _context2.next = 4;
                  break;
                }

                throw new (_BuildError || _load_BuildError()).default('Your project must have an Android package set in app.json\nSee https://docs.expo.io/versions/latest/distribution/building-standalone-apps/#2-configure-appjson');

              case 4:
                androidPackage = (0, (_get || _load_get()).default)(this.manifest, 'android.package');

                if (androidPackage) {
                  _context2.next = 7;
                  break;
                }

                throw new (_BuildError || _load_BuildError()).default('Your project must have an Android package set in app.json.');

              case 7:
                if (/^[a-zA-Z][a-zA-Z0-9_]*(\.[a-zA-Z][a-zA-Z0-9_]*)+$/.test(androidPackage)) {
                  _context2.next = 9;
                  break;
                }

                throw new (_BuildError || _load_BuildError()).default("Invalid format of Android package name (only alphanumeric characters, '.' and '_' are allowed, and each '.' must be followed by a letter)");

              case 9:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function validateProject() {
        return _ref2.apply(this, arguments);
      }

      return validateProject;
    }()
  }, {
    key: '_clearCredentials',
    value: function () {
      var _ref3 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3() {
        var username, experienceName, credentialMetadata, questions, answers, backupKeystoreOutputPath;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                username = this.user.username;
                experienceName = '@' + username + '/' + this.manifest.slug;
                credentialMetadata = {
                  username: username,
                  experienceName: experienceName,
                  platform: ANDROID
                };


                (_log || _load_log()).default.warn('Clearing your Android build credentials from our build servers is a ' + (_chalk || _load_chalk()).default.red('PERMANENT and IRREVERSIBLE action.'));
                (_log || _load_log()).default.warn('Android keystores must be identical to the one previously used to submit your app to the Google Play Store.');
                (_log || _load_log()).default.warn('Please read https://docs.expo.io/versions/latest/distribution/building-standalone-apps/#if-you-choose-to-build-for-android for more info before proceeding.');
                (_log || _load_log()).default.warn("We'll store a backup of your Android keystore in this directory in case you decide to delete it from our servers.");
                questions = [{
                  type: 'confirm',
                  name: 'confirm',
                  message: 'Permanently delete the Android build credentials from our servers?'
                }];
                _context3.next = 10;
                return (0, (_prompt || _load_prompt()).default)(questions);

              case 10:
                answers = _context3.sent;

                if (!answers.confirm) {
                  _context3.next = 19;
                  break;
                }

                (0, (_log || _load_log()).default)('Backing up your Android keystore now...');
                backupKeystoreOutputPath = _path.default.resolve(this.projectDir, this.manifest.slug + '.jks.bak');
                _context3.next = 16;
                return (_xdl || _load_xdl()).Credentials.Android.backupExistingCredentials({
                  outputPath: backupKeystoreOutputPath,
                  username: username,
                  experienceName: experienceName
                }, (_log || _load_log()).default);

              case 16:
                _context3.next = 18;
                return (_xdl || _load_xdl()).Credentials.removeCredentialsForPlatform(ANDROID, credentialMetadata);

              case 18:
                (_log || _load_log()).default.warn('Removed existing credentials from Expo servers');

              case 19:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function _clearCredentials() {
        return _ref3.apply(this, arguments);
      }

      return _clearCredentials;
    }()
  }, {
    key: 'collectAndValidateCredentials',
    value: function () {
      var _ref4 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee5() {
        var _this2 = this;

        var username, experienceName, credentialMetadata, credentialsExist, questions, answers, keystorePath, keystoreAlias, keystorePassword, keyPassword, keystoreData, credentials;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                username = this.user.username;
                experienceName = '@' + username + '/' + this.manifest.slug;
                credentialMetadata = {
                  username: username,
                  experienceName: experienceName,
                  platform: ANDROID
                };
                _context5.next = 5;
                return (_xdl || _load_xdl()).Credentials.credentialsExistForPlatformAsync(credentialMetadata);

              case 5:
                credentialsExist = _context5.sent;

                if (!this.checkEnv()) {
                  _context5.next = 11;
                  break;
                }

                _context5.next = 9;
                return this.collectAndValidateCredentialsFromCI(credentialMetadata);

              case 9:
                _context5.next = 30;
                break;

              case 11:
                if (!(this.options.clearCredentials || !credentialsExist)) {
                  _context5.next = 30;
                  break;
                }

                console.log('');
                questions = [{
                  type: 'rawlist',
                  name: 'uploadKeystore',
                  message: 'Would you like to upload a keystore or have us generate one for you?\nIf you don\'t know what this means, let us handle it! :)\n',
                  choices: [{ name: 'Let Expo handle the process!', value: false }, { name: 'I want to upload my own keystore!', value: true }]
                }, {
                  type: 'input',
                  name: 'keystorePath',
                  message: 'Path to keystore:',
                  validate: function () {
                    var _ref5 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee4(keystorePath) {
                      var keystorePathStats;
                      return (_regenerator || _load_regenerator()).default.wrap(function _callee4$(_context4) {
                        while (1) {
                          switch (_context4.prev = _context4.next) {
                            case 0:
                              _context4.prev = 0;
                              _context4.next = 3;
                              return (_fsExtra || _load_fsExtra()).default.stat(keystorePath);

                            case 3:
                              keystorePathStats = _context4.sent;
                              return _context4.abrupt('return', keystorePathStats.isFile());

                            case 7:
                              _context4.prev = 7;
                              _context4.t0 = _context4['catch'](0);

                              // file does not exist
                              console.log('\nFile does not exist.');
                              return _context4.abrupt('return', false);

                            case 11:
                            case 'end':
                              return _context4.stop();
                          }
                        }
                      }, _callee4, _this2, [[0, 7]]);
                    }));

                    return function validate(_x) {
                      return _ref5.apply(this, arguments);
                    };
                  }(),
                  filter: function filter(keystorePath) {
                    keystorePath = (0, (_untildify || _load_untildify()).default)(keystorePath);
                    if (!_path.default.isAbsolute(keystorePath)) {
                      keystorePath = _path.default.resolve(keystorePath);
                    }
                    return keystorePath;
                  },
                  when: function when(answers) {
                    return answers.uploadKeystore;
                  }
                }, {
                  type: 'input',
                  name: 'keystoreAlias',
                  message: 'Keystore Alias:',
                  validate: function validate(val) {
                    return val !== '';
                  },
                  when: function when(answers) {
                    return answers.uploadKeystore;
                  }
                }, {
                  type: 'password',
                  name: 'keystorePassword',
                  message: 'Keystore Password:',
                  validate: function validate(val) {
                    return val !== '';
                  },
                  when: function when(answers) {
                    return answers.uploadKeystore;
                  }
                }, {
                  type: 'password',
                  name: 'keyPassword',
                  message: 'Key Password:',
                  validate: function validate(password, answers) {
                    if (password === '') {
                      return false;
                    }
                    // Todo validate keystore passwords
                    return true;
                  },
                  when: function when(answers) {
                    return answers.uploadKeystore;
                  }
                }];
                _context5.next = 16;
                return (0, (_prompt || _load_prompt()).default)(questions);

              case 16:
                answers = _context5.sent;

                if (answers.uploadKeystore) {
                  _context5.next = 23;
                  break;
                }

                if (!(this.options.clearCredentials && credentialsExist)) {
                  _context5.next = 21;
                  break;
                }

                _context5.next = 21;
                return this._clearCredentials();

              case 21:
                _context5.next = 30;
                break;

              case 23:
                keystorePath = answers.keystorePath, keystoreAlias = answers.keystoreAlias, keystorePassword = answers.keystorePassword, keyPassword = answers.keyPassword;

                // read the keystore

                _context5.next = 26;
                return (_fsExtra || _load_fsExtra()).default.readFile(keystorePath);

              case 26:
                keystoreData = _context5.sent;
                credentials = {
                  keystore: keystoreData.toString('base64'),
                  keystoreAlias: keystoreAlias,
                  keystorePassword: keystorePassword,
                  keyPassword: keyPassword
                };
                _context5.next = 30;
                return (_xdl || _load_xdl()).Credentials.updateCredentialsForPlatform(ANDROID, credentials, [], credentialMetadata);

              case 30:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function collectAndValidateCredentials() {
        return _ref4.apply(this, arguments);
      }

      return collectAndValidateCredentials;
    }()
  }, {
    key: 'checkEnv',
    value: function checkEnv() {
      return this.options.keystorePath && this.options.keystoreAlias && process.env.EXPO_ANDROID_KEYSTORE_PASSWORD && process.env.EXPO_ANDROID_KEY_PASSWORD;
    }
  }, {
    key: 'collectAndValidateCredentialsFromCI',
    value: function () {
      var _ref6 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee6(credentialMetadata) {
        var credentials;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return (_fsExtra || _load_fsExtra()).default.readFile(this.options.keystorePath);

              case 2:
                _context6.t0 = _context6.sent.toString('base64');
                _context6.t1 = this.options.keystoreAlias;
                _context6.t2 = process.env.EXPO_ANDROID_KEYSTORE_PASSWORD;
                _context6.t3 = process.env.EXPO_ANDROID_KEY_PASSWORD;
                credentials = {
                  keystore: _context6.t0,
                  keystoreAlias: _context6.t1,
                  keystorePassword: _context6.t2,
                  keyPassword: _context6.t3
                };
                _context6.next = 9;
                return (_xdl || _load_xdl()).Credentials.updateCredentialsForPlatform(ANDROID, credentials, [], credentialMetadata);

              case 9:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function collectAndValidateCredentialsFromCI(_x2) {
        return _ref6.apply(this, arguments);
      }

      return collectAndValidateCredentialsFromCI;
    }()
  }, {
    key: 'platform',
    value: function platform() {
      return ANDROID;
    }
  }]);
  return AndroidBuilder;
}((_BaseBuilder2 || _load_BaseBuilder()).default);

exports.default = AndroidBuilder;
module.exports = exports['default'];
