// expo-cli/build/commands/build/BaseBuilder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2;

function _load_extends() {
  return _extends2 = _interopRequireDefault(require('babel-runtime/helpers/extends'));
}

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var _classCallCheck2;

function _load_classCallCheck() {
  return _classCallCheck2 = _interopRequireDefault(require('babel-runtime/helpers/classCallCheck'));
}

var _createClass2;

function _load_createClass() {
  return _createClass2 = _interopRequireDefault(require('babel-runtime/helpers/createClass'));
}

var _promise;

function _load_promise() {
  return _promise = _interopRequireDefault(require('babel-runtime/core-js/promise'));
}

var _xdl;

function _load_xdl() {
  return _xdl = require('@expo/xdl');
}

var _chalk;

function _load_chalk() {
  return _chalk = _interopRequireDefault(require('chalk'));
}

var _fp;

function _load_fp() {
  return _fp = _interopRequireDefault(require('lodash/fp'));
}

var _get;

function _load_get() {
  return _get = _interopRequireDefault(require('lodash/get'));
}

var _ora;

function _load_ora() {
  return _ora = _interopRequireDefault(require('ora'));
}

var _semver;

function _load_semver() {
  return _semver = _interopRequireDefault(require('semver'));
}

var _url;

function _load_url() {
  return _url = _interopRequireWildcard(require('../utils/url'));
}

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('../../log'));
}

var _publish;

function _load_publish() {
  return _publish = require('../publish');
}

var _BuildError;

function _load_BuildError() {
  return _BuildError = _interopRequireDefault(require('./BuildError'));
}

var _prompt;

function _load_prompt() {
  return _prompt = _interopRequireDefault(require('../../prompt'));
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sleep = function sleep(ms) {
  return new (_promise || _load_promise()).default(function (res) {
    return setTimeout(res, ms);
  });
};

var secondsToMilliseconds = function secondsToMilliseconds(seconds) {
  return seconds * 1000;
};

var BaseBuilder = function () {
  function BaseBuilder(projectDir) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    (0, (_classCallCheck2 || _load_classCallCheck()).default)(this, BaseBuilder);
    this.projectDir = '';
    this.options = {
      wait: true,
      clearCredentials: false,
      releaseChannel: 'default',
      publish: false
    };
    this.manifest = {};

    this.projectDir = projectDir;
    this.options = options;
  }

  (0, (_createClass2 || _load_createClass()).default)(BaseBuilder, [{
    key: 'command',
    value: function () {
      var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee() {
        return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.prev = 0;
                _context.next = 3;
                return this.prepareProjectInfo();

              case 3:
                _context.next = 5;
                return this.run();

              case 5:
                _context.next = 15;
                break;

              case 7:
                _context.prev = 7;
                _context.t0 = _context['catch'](0);

                if (_context.t0 instanceof (_BuildError || _load_BuildError()).default) {
                  _context.next = 13;
                  break;
                }

                throw _context.t0;

              case 13:
                (_log || _load_log()).default.error(_context.t0.message);
                process.exit(1);

              case 15:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this, [[0, 7]]);
      }));

      function command() {
        return _ref.apply(this, arguments);
      }

      return command;
    }()
  }, {
    key: 'commandCheckStatus',
    value: function () {
      var _ref2 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2() {
        return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.prev = 0;
                _context2.next = 3;
                return this.prepareProjectInfo();

              case 3:
                _context2.next = 5;
                return this.checkStatus();

              case 5:
                _context2.next = 15;
                break;

              case 7:
                _context2.prev = 7;
                _context2.t0 = _context2['catch'](0);

                if (_context2.t0 instanceof (_BuildError || _load_BuildError()).default) {
                  _context2.next = 13;
                  break;
                }

                throw _context2.t0;

              case 13:
                (_log || _load_log()).default.error(_context2.t0.message);
                process.exit(1);

              case 15:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this, [[0, 7]]);
      }));

      function commandCheckStatus() {
        return _ref2.apply(this, arguments);
      }

      return commandCheckStatus;
    }()
  }, {
    key: 'prepareProjectInfo',
    value: function () {
      var _ref3 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3() {
        var _ref4, exp;

        return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return (_xdl || _load_xdl()).ProjectUtils.readConfigJsonAsync(this.projectDir);

              case 2:
                _ref4 = _context3.sent;
                exp = _ref4.exp;

                this.manifest = exp;
                _context3.next = 7;
                return (_xdl || _load_xdl()).User.ensureLoggedInAsync();

              case 7:
                this.user = _context3.sent;
                _context3.next = 10;
                return this.checkProjectConfig();

              case 10:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function prepareProjectInfo() {
        return _ref3.apply(this, arguments);
      }

      return prepareProjectInfo;
    }()
  }, {
    key: 'checkProjectConfig',
    value: function () {
      var _ref5 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee4() {
        var oldestSupportedMajorVersion, _ref6, version;

        return (_regenerator || _load_regenerator()).default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                if (this.manifest.isDetached) {
                  (_log || _load_log()).default.error('\'expo build:' + this.platform() + '\' is not supported for detached projects.');
                  process.exit(1);
                }

                // Warn user if building a project using the next deprecated SDK version
                _context4.next = 3;
                return (_xdl || _load_xdl()).Versions.oldestSupportedMajorVersionAsync();

              case 3:
                oldestSupportedMajorVersion = _context4.sent;

                if (!((_semver || _load_semver()).default.major(this.manifest.sdkVersion) === oldestSupportedMajorVersion)) {
                  _context4.next = 10;
                  break;
                }

                _context4.next = 7;
                return (_xdl || _load_xdl()).Versions.newestSdkVersionAsync();

              case 7:
                _ref6 = _context4.sent;
                version = _ref6.version;

                (_log || _load_log()).default.warn('\nSDK' + oldestSupportedMajorVersion + ' will be ' + (_chalk || _load_chalk()).default.bold('deprecated') + ' next! We recommend upgrading versions, ideally to the latest (SDK' + (_semver || _load_semver()).default.major(version) + '), so you can continue to build new binaries of your app and develop in the Expo Client.\n');

              case 10:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function checkProjectConfig() {
        return _ref5.apply(this, arguments);
      }

      return checkProjectConfig;
    }()
  }, {
    key: 'checkForBuildInProgress',
    value: function () {
      var _ref7 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee5() {
        var buildStatus;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                (0, (_log || _load_log()).default)('Checking if there is a build in progress...\n');
                _context5.next = 3;
                return (_xdl || _load_xdl()).Project.buildAsync(this.projectDir, {
                  mode: 'status',
                  platform: this.platform(),
                  current: true,
                  releaseChannel: this.options.releaseChannel,
                  publicUrl: this.options.publicUrl,
                  sdkVersion: this.manifest.sdkVersion
                });

              case 3:
                buildStatus = _context5.sent;

                if (!(buildStatus.jobs && buildStatus.jobs.length)) {
                  _context5.next = 6;
                  break;
                }

                throw new (_BuildError || _load_BuildError()).default('Cannot start a new build, as there is already an in-progress build.');

              case 6:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function checkForBuildInProgress() {
        return _ref7.apply(this, arguments);
      }

      return checkForBuildInProgress;
    }()
  }, {
    key: 'checkStatus',
    value: function () {
      var _ref8 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee6() {
        var platform = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'all';
        var buildStatus;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                (0, (_log || _load_log()).default)('Fetching build history...\n');
                _context6.next = 3;
                return (_xdl || _load_xdl()).Project.buildAsync(this.projectDir, {
                  mode: 'status',
                  platform: platform,
                  current: false,
                  releaseChannel: this.options.releaseChannel
                });

              case 3:
                buildStatus = _context6.sent;

                if (!buildStatus.err) {
                  _context6.next = 6;
                  break;
                }

                throw new Error('Error getting current build status for this project.');

              case 6:
                if (buildStatus.jobs && buildStatus.jobs.length) {
                  _context6.next = 9;
                  break;
                }

                (0, (_log || _load_log()).default)('No currently active or previous builds for this project.');
                return _context6.abrupt('return');

              case 9:

                this.logBuildStatuses(buildStatus);

              case 10:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function checkStatus() {
        return _ref8.apply(this, arguments);
      }

      return checkStatus;
    }()
  }, {
    key: 'checkStatusBeforeBuild',
    value: function () {
      var _ref9 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee7() {
        var reuseStatus, questions, answers;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                (0, (_log || _load_log()).default)('Checking if this build already exists...\n');

                _context7.next = 3;
                return (_xdl || _load_xdl()).Project.findReusableBuildAsync(this.options.releaseChannel, this.platform(), this.manifest.sdkVersion, this.manifest.slug);

              case 3:
                reuseStatus = _context7.sent;

                if (!reuseStatus.canReuse) {
                  _context7.next = 12;
                  break;
                }

                (_log || _load_log()).default.warn('Did you know that Expo provides over-the-air updates?\nPlease see the docs (' + (_chalk || _load_chalk()).default.underline('https://docs.expo.io/versions/latest/guides/configuring-ota-updates/') + ') and check if you can use them instead of building your app binaries again.');

                (_log || _load_log()).default.warn('There were no new changes from the last build, you can download that build from here: ' + (_chalk || _load_chalk()).default.underline(reuseStatus.downloadUrl));

                questions = [{
                  type: 'confirm',
                  name: 'confirm',
                  message: 'Do you want to build app anyway?'
                }];
                _context7.next = 10;
                return (0, (_prompt || _load_prompt()).default)(questions);

              case 10:
                answers = _context7.sent;

                if (!answers.confirm) {
                  (0, (_log || _load_log()).default)('Stopping the build process');
                  process.exit(0);
                }

              case 12:
              case 'end':
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function checkStatusBeforeBuild() {
        return _ref9.apply(this, arguments);
      }

      return checkStatusBeforeBuild;
    }()
  }, {
    key: 'logBuildStatuses',
    value: function logBuildStatuses(buildStatus) {
      (_log || _load_log()).default.raw();
      (0, (_log || _load_log()).default)('=================');
      (0, (_log || _load_log()).default)(' Builds Statuses ');
      (0, (_log || _load_log()).default)('=================\n');
      buildStatus.jobs.forEach(function (job, i) {
        var platform = void 0,
            packageExtension = void 0;
        if (job.platform === 'ios') {
          platform = 'iOS';
          packageExtension = 'IPA';
        } else {
          platform = 'Android';
          packageExtension = 'APK';
        }

        (0, (_log || _load_log()).default)('### ' + i + ' | ' + platform + ' | ' + (_url || _load_url()).constructBuildLogsUrl(job.id) + ' ###');

        var hasPriorityBuilds = buildStatus.numberOfRemainingPriorityBuilds > 0 || buildStatus.hasUnlimitedPriorityBuilds;
        var shouldShowUpgradeInfo = !hasPriorityBuilds && i === 0 && job.priority === 'normal' && buildStatus.canPurchasePriorityBuilds;
        var status = void 0;
        switch (job.status) {
          case 'pending':
          case 'sent-to-queue':
            status = 'Build waiting in queue...\nQueue length: ' + (_chalk || _load_chalk()).default.underline((_url || _load_url()).constructTurtleStatusUrl());
            if (shouldShowUpgradeInfo) {
              status += '\nWant to wait less? Get priority builds at ' + (_chalk || _load_chalk()).default.underline('https://expo.io/settings/billing') + '.';
            }
            break;
          case 'started':
            status = 'Build started...';
            break;
          case 'in-progress':
            status = 'Build in progress...';
            if (shouldShowUpgradeInfo) {
              status += '\nWant to wait less? Get priority builds at ' + (_chalk || _load_chalk()).default.underline('https://expo.io/settings/billing') + '.';
            }
            break;
          case 'finished':
            status = 'Build finished.';
            if (shouldShowUpgradeInfo) {
              status += '\nLooks like this build could have been faster.\nRead more about priority builds at ' + (_chalk || _load_chalk()).default.underline('https://expo.io/settings/billing') + '.';
            }
            break;
          case 'errored':
            status = 'There was an error with this build.';
            if (job.id) {
              status += '\n\nWhen requesting support, please provide this build ID:\n\n' + job.id + '\n';
            }
            break;
          default:
            status = '';
            break;
        }

        (0, (_log || _load_log()).default)(status);
        if (job.status === 'finished') {
          if (job.artifacts) {
            (0, (_log || _load_log()).default)(packageExtension + ': ' + job.artifacts.url);
          } else {
            (0, (_log || _load_log()).default)('Problem getting ' + packageExtension + ' URL. Please try to build again.');
          }
        }
        (0, (_log || _load_log()).default)();
      });
    }
  }, {
    key: 'ensureReleaseExists',
    value: function () {
      var _ref10 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee8() {
        var _ref11, ids, url, err, release;

        return (_regenerator || _load_regenerator()).default.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                if (!this.options.publish) {
                  _context8.next = 16;
                  break;
                }

                _context8.next = 3;
                return (0, (_publish || _load_publish()).action)(this.projectDir, (0, (_extends2 || _load_extends()).default)({}, this.options, {
                  platform: this.platform(),
                  duringBuild: true
                }));

              case 3:
                _ref11 = _context8.sent;
                ids = _ref11.ids;
                url = _ref11.url;
                err = _ref11.err;

                if (!err) {
                  _context8.next = 11;
                  break;
                }

                throw new (_BuildError || _load_BuildError()).default('No url was returned from publish. Please try again.\n' + err);

              case 11:
                if (!(!url || url === '')) {
                  _context8.next = 13;
                  break;
                }

                throw new (_BuildError || _load_BuildError()).default('No url was returned from publish. Please try again.');

              case 13:
                return _context8.abrupt('return', ids);

              case 16:
                (0, (_log || _load_log()).default)('Looking for releases...');
                _context8.next = 19;
                return (_xdl || _load_xdl()).Project.getLatestReleaseAsync(this.projectDir, {
                  releaseChannel: this.options.releaseChannel,
                  platform: this.platform()
                });

              case 19:
                release = _context8.sent;

                if (release) {
                  _context8.next = 22;
                  break;
                }

                throw new (_BuildError || _load_BuildError()).default('No releases found. Please create one using `exp publish` first.');

              case 22:
                (0, (_log || _load_log()).default)('Using existing release on channel "' + release.channel + '":\n' + ('publicationId: ' + release.publicationId + '\n  publishedTime: ' + release.publishedTime));
                return _context8.abrupt('return', [release.publicationId]);

              case 24:
              case 'end':
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function ensureReleaseExists() {
        return _ref10.apply(this, arguments);
      }

      return ensureReleaseExists;
    }()
  }, {
    key: 'wait',
    value: function () {
      var _ref12 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee9(buildId) {
        var _ref13 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
            _ref13$timeout = _ref13.timeout,
            timeout = _ref13$timeout === undefined ? 1200 : _ref13$timeout,
            _ref13$interval = _ref13.interval,
            interval = _ref13$interval === undefined ? 30 : _ref13$interval,
            publicUrl = _ref13.publicUrl;

        var spinner, time, endTime, res, job;
        return (_regenerator || _load_regenerator()).default.wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                (0, (_log || _load_log()).default)('Waiting for build to complete. You can press Ctrl+C to exit.');
                spinner = (0, (_ora || _load_ora()).default)().start();
                time = new Date().getTime();
                endTime = time + secondsToMilliseconds(timeout);

              case 4:
                if (!(time <= endTime)) {
                  _context9.next = 27;
                  break;
                }

                _context9.next = 7;
                return (_xdl || _load_xdl()).Project.buildAsync(this.projectDir, (0, (_extends2 || _load_extends()).default)({
                  current: false,
                  mode: 'status'
                }, publicUrl ? { publicUrl: publicUrl } : {}));

              case 7:
                res = _context9.sent;
                job = (_fp || _load_fp()).default.compose((_fp || _load_fp()).default.head, (_fp || _load_fp()).default.filter(function (job) {
                  return buildId && job.id === buildId;
                }), (_fp || _load_fp()).default.getOr([], 'jobs'))(res);
                _context9.t0 = job.status;
                _context9.next = _context9.t0 === 'finished' ? 12 : _context9.t0 === 'pending' ? 14 : _context9.t0 === 'sent-to-queue' ? 14 : _context9.t0 === 'started' ? 16 : _context9.t0 === 'in-progress' ? 16 : _context9.t0 === 'errored' ? 18 : 20;
                break;

              case 12:
                spinner.succeed('Build finished.');
                return _context9.abrupt('return', job);

              case 14:
                spinner.text = 'Build queued...';
                return _context9.abrupt('break', 22);

              case 16:
                spinner.text = 'Build in progress...';
                return _context9.abrupt('break', 22);

              case 18:
                spinner.fail('Build failed.');
                throw new (_BuildError || _load_BuildError()).default('Standalone build failed!');

              case 20:
                spinner.warn('Unknown status.');
                throw new (_BuildError || _load_BuildError()).default('Unknown status: ' + job.status + ' - aborting!');

              case 22:
                time = new Date().getTime();
                _context9.next = 25;
                return sleep(secondsToMilliseconds(interval));

              case 25:
                _context9.next = 4;
                break;

              case 27:
                spinner.warn('Timed out.');
                throw new (_BuildError || _load_BuildError()).default('Timeout reached! Project is taking longer than expected to finish building, aborting wait...');

              case 29:
              case 'end':
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function wait(_x3) {
        return _ref12.apply(this, arguments);
      }

      return wait;
    }()
  }, {
    key: 'build',
    value: function () {
      var _ref14 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee10(expIds) {
        var publicUrl, platform, bundleIdentifier, opts, _ref15, buildId, priority, canPurchasePriorityBuilds, waitOpts, completedJob, artifactUrl;

        return (_regenerator || _load_regenerator()).default.wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                publicUrl = this.options.publicUrl;
                platform = this.platform();
                bundleIdentifier = (0, (_get || _load_get()).default)(this.manifest, 'ios.bundleIdentifier');
                opts = (0, (_extends2 || _load_extends()).default)({
                  mode: 'create',
                  expIds: expIds,
                  platform: platform,
                  releaseChannel: this.options.releaseChannel
                }, publicUrl ? { publicUrl: publicUrl } : {});


                if (platform === 'ios') {
                  opts = (0, (_extends2 || _load_extends()).default)({}, opts, {
                    type: this.options.type,
                    bundleIdentifier: bundleIdentifier
                  });
                } else if (platform === 'android') {
                  opts = (0, (_extends2 || _load_extends()).default)({}, opts, {
                    type: this.options.type
                  });
                }

                // call out to build api here with url
                _context10.next = 7;
                return (_xdl || _load_xdl()).Project.buildAsync(this.projectDir, opts);

              case 7:
                _ref15 = _context10.sent;
                buildId = _ref15.id;
                priority = _ref15.priority;
                canPurchasePriorityBuilds = _ref15.canPurchasePriorityBuilds;


                (0, (_log || _load_log()).default)('Build started, it may take a few minutes to complete.');
                (0, (_log || _load_log()).default)('You can check the queue length at ' + (_chalk || _load_chalk()).default.underline((_url || _load_url()).constructTurtleStatusUrl()) + '\n');
                if (priority === 'normal' && canPurchasePriorityBuilds) {
                  (0, (_log || _load_log()).default)('You can make this faster. ðŸ¢\nGet priority builds at: https://expo.io/settings/billing\n');
                }

                if (buildId) {
                  (0, (_log || _load_log()).default)('You can monitor the build at\n\n ' + (_chalk || _load_chalk()).default.underline((_url || _load_url()).constructBuildLogsUrl(buildId)) + '\n');
                }

                if (!this.options.wait) {
                  _context10.next = 24;
                  break;
                }

                waitOpts = publicUrl ? { publicUrl: publicUrl } : {};
                _context10.next = 19;
                return this.wait(buildId, waitOpts);

              case 19:
                completedJob = _context10.sent;
                artifactUrl = completedJob.artifactId ? (_url || _load_url()).constructArtifactUrl(completedJob.artifactId) : completedJob.artifacts.url;

                (0, (_log || _load_log()).default)((_chalk || _load_chalk()).default.green('Successfully built standalone app:') + ' ' + (_chalk || _load_chalk()).default.underline(artifactUrl));
                _context10.next = 25;
                break;

              case 24:
                (0, (_log || _load_log()).default)('Alternatively, run `exp build:status` to monitor it from the command line.');

              case 25:
              case 'end':
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function build(_x5) {
        return _ref14.apply(this, arguments);
      }

      return build;
    }()
  }, {
    key: 'platform',
    value: function platform() {
      return 'all';
    }
  }]);
  return BaseBuilder;
}();

exports.default = BaseBuilder;
module.exports = exports['default'];
