// @expo/xdl/build/Sentry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.logError = logError;
exports.captureException = captureException;

var _raven = _interopRequireDefault(require("raven"));

var _Config = _interopRequireDefault(require("./Config"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const SENTRY_DSN = 'https://8554f14d112d4ed4b0558154762760ef:bae5673d5e5243abac5563d70861b5d8@sentry.io/194120';

_raven.default.config(SENTRY_DSN).install();

function logError(message, options) {
  // send error to Sentry
  // add `testing: true` to tags to avoid sending an email when testing
  _raven.default.captureMessage(message, getOptions(options));
}

function captureException(ex, options) {
  _raven.default.captureException(ex, getOptions(options));
}

function getOptions(options = {}) {
  return { ...options,
    tags: { ...options.tags,
      developerTool: _Config.default.developerTool,
      offline: _Config.default.offline
    }
  };
}
